### PR TITLE
Additional feature enhancements to nxos_logging

### PR DESCRIPTION
**Pull request by tstoner**
##### SUMMARY

Additional feature enhancements to nxos_logging plus module context infrastructure.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

nxos_logging

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (nxos_logging 14c61e8fb1) last updated 2018/09/19 09:24:54 (GMT -400)
  config file = None
  configured module search path = [u'/Users/tmstoner/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/tmstoner/ansible/lib/ansible
  executable location = /Users/tmstoner/ansible/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### ADDITIONAL INFORMATION

Added additional configuration support for nxos_logging.  Infrastructure was added to nxos.py plug in to allow modules to cache persistent context over lifetime of the playbook.  The intent of this infrastructure is to allow nxos modules to build up state over the lifetime of the play to such that unwanted configuration or configuration that does not match the playbook can be purged from the system.

```paste below

```

### CODE REVIEW

File: lib/ansible/module_utils/network/nxos/nxos.py [hunk 1]
```diff
@@ -464,6 +483,18 @@ def get_capabilities(self):
         result['network_api'] = 'nxapi'
         return result
 
+    def read_module_context(self, module_key):
+        Connection('')
```
Comment (mikewiebe) [inline lib/ansible/module_utils/network/nxos/nxos.py]: Why is this call to `Connection('')` needed?

File: lib/ansible/modules/network/nxos/nxos_logging.py [hunk 2]
```diff
@@ -69,6 +69,36 @@
...
     choices: ['present', 'absent']
+  event:
+    description:
+      - Link/trunk enable/default interface configuration logging
+    choices: ['link-enable', 'link-default', 'trunk-enable', 'trunk-default']
+    version_added: '2.8'
+  message:
+    description:
+      - Add interface description to interface syslogs [default: ethernet]
+      - Does not work with 6.0 images using nxapi as a transport.
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_logging.py]: *nit* `version 6.0 images`

**Comment by mikewiebe** [comment]
@tstoner Update summary to indicate `nxos_logging` instead of `nxapi_logging`

File: lib/ansible/modules/network/nxos/nxos_logging.py [hunk 3]
```diff
@@ -69,6 +69,36 @@
...
+    choices: ['add-interface-description']
+    version_added: '2.8'
+  file_size:
+    description:
+      - Set logfile size
+    version_added: '2.8'
+  facility_link_status:
+    description:
+      - Set logging facility ethpm link status.  Not idempotent with 6.0 images.
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_logging.py]: *nit* `version 6.0 images`

File: lib/ansible/modules/network/nxos/nxos_logging.py [hunk 4]
```diff
@@ -111,7 +147,26 @@
   nxos_logging:
     interface: mgmt0
     state: present
-
+- name: Purge configuration not what was just configured
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_logging.py]: I found this description a bit confusing.  Perhaps the following wording instead?

`- name: Purge nxos_logging configuration not managed by this playbook`

File: lib/ansible/plugins/cliconf/nxos.py [hunk 5]
```diff
@@ -36,8 +36,21 @@
 class Cliconf(CliconfBase):
 
     def __init__(self, *args, **kwargs):
+        self._module_context = {}
         super(Cliconf, self).__init__(*args, **kwargs)
 
+    def read_module_context(self, module_key):
+        #return [{'logging': 'this is a test'}]
```
Comment (mikewiebe) [inline lib/ansible/plugins/cliconf/nxos.py]: Remove?

File: lib/ansible/module_utils/network/nxos/nxos.py (hunk 1 above)
Comment (tstoner) [reply lib/ansible/module_utils/network/nxos/nxos.py]: Good catch.

File: lib/ansible/modules/network/nxos/nxos_logging.py [hunk 6]
```diff
@@ -458,8 +724,14 @@ def main():
...
         interface=dict(),
+        facility_link_status=dict(choices=['link-down-notif', 'link-down-error', 'link-up-notif', 'link-up-error']),
+        event=dict(choices=['link-enable', 'link-default', 'trunk-enable', 'trunk-default']),
+        message=dict(choices=['add-interface-description']),
+        file_size=dict(type='int'),
+        timestamp=dict(choices=['microseconds', 'milliseconds', 'seconds']),
         state=dict(default='present', choices=['present', 'absent']),
-        aggregate=dict(type='list')
+        aggregate=dict(type='list'),
+        purge=dict(default=False, type='bool')
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_logging.py:736]: I remember we discussed adding the `purge` option to the default `nxos_argument_spec`.  Is this something we should do now?

Comment (tstoner) [reply lib/ansible/modules/network/nxos/nxos_logging.py:736]: This is where I had it initially but decided to make it module specific for now.  I think when we start implementing purge for other nxos modules we can add it to the nxos_argument_spec.  IMO.

**Review by trishnaguha — APPROVED** [review]

**Pull request merged**